### PR TITLE
C++: More effective barriers in the `bounded` predicate for CWE-190

### PR DESCRIPTION
**Pull request by MathiasVP**
<s><b>Note</b>: This PR is based on top of https://github.com/github/codeql/pull/6154. Please start the review at 142b78fe5a1ccdf308fff9b65220f0407838360b. Once #6154 has been merged I'll rebase those commits away.</s> <-- That PR is now merged, and has been rebased out of this PR.

---

This PR makes `%` a barrier regardless of whether or not we can bound the right-hand side of the operator using range analysis.

The barrier (with that restriction) was introduced in https://github.com/github/codeql/pull/5887. At that point, we didn't have a lot of results on the query since it only recognized calls to `rand()` as random sources.

Now that we have more sources of randomness (i.e., since https://github.com/github/codeql/pull/6154) we have a lot more results, and it looks like the `%` barrier doesn't rule out as many results as we want it to. For examples of this, see the first two bullet points I wrote here: https://github.com/github/codeql-c-team/issues/553#issuecomment-867670535.

Here's a difference run on our usual LGTM projects: https://lgtm.com/query/2650148462537684414/. All of the removed results look like false positives to me. It also removes some false positives on the `cpp/tainted-arithmetic` query, but it's difficult to create an LGTM difference query since `cpp/tainted-arithmetic` is still using the `TaintTrackingConfiguration` configuration.

This change has no effect on SAMATE.

### CODE REVIEW

**Comment by rdmarsh2** [comment]
LGTM once  #6154 is merged and the rebase is done.

**Comment by MathiasVP** [comment]
@rdmarsh2 https://github.com/github/codeql/pull/6154 is now merged and has been rebased out of this PR. I think this one should now be good to go.

**Comment by rdmarsh2** [comment]
Is there a reason to keep `boundedRem` rather than just writing `e = any(RemExpr r).getLeftOperand()`?

**Comment by MathiasVP** [comment]
> Is there a reason to keep `boundedRem` rather than just writing `e = any(RemExpr r).getLeftOperand()`?

Not really, no. I suppose we could inline them all at this point.

**Review by rdmarsh2 — APPROVED** [review]

**Pull request merged**
